File: backend/database/passwords.py
import random
import string
import csv


def generate_passwords(num_passwords, max_chars=24) -> set[str]:
    """Generate a set of passwords containing at least one lowercase letter,
    at least one uppercase letter, at least one digit, and one special character.

    :param num_passwords: the number of passwords generated
    :param max_chars: the maximum number of characters in a password
    :return: the set of generated passwords
    """
    MIN_CHARS = 10

    passwords = set()
    while len(passwords) < num_passwords:
        # select one lowercase
        password = random.choice(string.ascii_lowercase)
        # select one uppercase
        password += random.choice(string.ascii_uppercase)
        # select one digit
        password += random.choice(string.digits)
        # select one special symbol
        password += random.choice(string.punctuation)

        # generate other characters
        character_options = string.ascii_letters + string.digits
        length = random.randint(MIN_CHARS - len(password), max_chars - len(password))
        for _ in range(length):
            password += random.choice(character_options)

        # shuffle all characters
        password = "".join(random.shuffle(list(password)))

        if password not in passwords:
            passwords.add(password)

    return passwords


# def write_passwords(file_path, passwords):
#     with open(file_path, 'w', newline="") as file:

File: backend/src/passwords.cpp
#include <iostream>
#include "passwords.hpp"

namespace passwords
{
    std::unordered_set<std::string> generatePasswords(const int &numPasswords, const int &maxChars)
    {
        int minChars = 10;

        if (maxChars < minChars) {
            std::cout << "Password must have a minimum length of 10 characters\n";
            exit(0);
        }

        std::unordered_set<std::string> passwords;
        std::string lowercase = "abcdefghijklmnopqrstuvwxyz";
        std::string uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        std::string digits = "0123456789";
        std::string symbols = "!#$%&'()\\*+,-./:;<=>?@[]^_`{|}~\"";

        while (passwords.size() < numPasswords)
        {
            // randomize length between minChars and maxChars
            int length = minChars + rand() % ((maxChars + 1) - minChars);
            std::string password = ""; 
            bool hasLowercase = false, hasUppercase = false, hasDigit = false, hasSymbol = false;

            for (int i = 0; i < length; i++)
            {
                int charType = rand() % 4;
                int index;
                switch (charType)
                {
                case 0: // lowercase letter
                    hasLowercase = true;
                    if (!(hasLowercase && hasUppercase && hasDigit && hasSymbol)){
                        break;
                    }
                    index = rand() % lowercase.length();
                    password = password + lowercase[index];
                    break;
                case 1: // uppercase letter
                    hasUppercase = true;
                    if (!(hasLowercase && hasUppercase && hasDigit && hasSymbol)) {
                        break;
                    }
                    index = rand() % uppercase.length();
                    password = password + uppercase[index];
                    break;
                case 2: // digit
                    hasDigit = true;
                    if (!(hasLowercase && hasUppercase && hasDigit && hasSymbol)) {
                        break;
                    }
                    index = rand() % digits.length();
                    password = password + digits[index];
                    break;
                case 3: // symbol
                    hasSymbol = true;
                    if (!(hasLowercase && hasUppercase && hasDigit && hasSymbol)) {
                        break;
                    }
                    index = rand() % symbols.length();
                    password = password + symbols[index];
                    break;
                };
            };
            passwords.insert(password);
        };
        return passwords;
    };
};